### PR TITLE
Add doc for fine-tuning.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-lash/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/general/finetuning.rst [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    trainer = flash.Trainer()
+    trainer.finetune(model, data, strategy=FreezeUnfreeze(unfreeze_epoch=5))
+
+Create a custom Finetuning Strategy
+
+.. code-block:: python
+
+    from flash.core.finetuning import BaseFinetuning
+
+    class FeatureExtractorFreezeUnfreeze(BaseFinetuning):
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:66]: subclass `FlashBaseFinetuning`?

File: docs/source/general/finetuning.rst [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+
+Finetune options
+================
+
+Flash finetune `strategy` argument can either a string or an instance of :class:`~python_lightning.callbacks.finetuning.BaseFinetuning`.
+Furthermore, Flash supports 4 builts-in Finetuning Callback accessible string
+
+    "no_freeze": BaseFinetuning,
+    "freeze": FlashBaseFinetuning,
```
Comment (carmocca) [inline docs/source/general/finetuning.rst]: Update these

File: docs/source/general/finetuning.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
+**********
+Finetuning
+**********
+
+Some Flash tasks have been pretrained on large data sets, to accelerate your training (calling the :func:`~flash.Trainer.finetune` method using a pretrained backbone will fine-tune the backbone to generate a model customized to your data set and desired task). If you want to train the task from scratch instead, pass `pretrained=False` parameter when creating your task. Then, use the :func:`~flash.Trainer.fit` method to train your model.
+When calling the :func:`~flash.Trainer.finetune`, one also needs to provide a finetune `strategy`.
```
Comment (SeanNaren) [inline docs/source/general/finetuning.rst:6]:
```suggestion
When training a Flash task, calling the :func:`~flash.Trainer.finetune` using a pre-trained backbone will finetune the backbone onto your data and desired task. If you would like to train from scratch, pass `pretrained=False` when creating your task whilst using the :func:`~flash.Trainer.fit` method to start training.

When using :func:`~flash.Trainer.finetune`, you also need to provide a finetune `strategy`.
```

File: docs/source/general/finetuning.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
+**********
+Finetuning
+**********
+
+When training a Flash task, calling the :func:`~flash.Trainer.finetune` using a pre-trained backbone will finetune the backbone onto your data and desired task. If you would like to train from scratch, pass `pretrained=False` when creating your task whilst using the :func:`~flash.Trainer.fit` method to start training.
```
Comment (edenlightning) [inline docs/source/general/finetuning.rst:5]:
```suggestion
Finetuning (or transfer-learning) is the process of tweaking a model trained on a large dataset, to your particular (likely much smaller) dataset. All Flash tasks have a pre-trained backbone that was trained on large datasets such as ImageNet, and that allows to decrease training time significantly. 

.. note:: If you have a huge dataset and prefer to train from scratch, see :ref:`training` guide.
```

File: docs/source/general/finetuning.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+Finetune options
+================
+
+Flash finetune `strategy` argument can either a string or an instance of :class:`~python_lightning.callbacks.finetuning.BaseFinetuning`.
+Furthermore, Flash supports 4 builts-in Finetuning Callback accessible via those strings:
+
+* `no_freeze`: Don't freeze anything.
+* `freeze`: Freeze the backbone parameters when training starts.
+* `freeze_unfreeze`: Freeze the backbone parameters when training starts and unfreeze the backbone when reaching `unfreeze_epoch`.
+* `unfreeze_milestones`: Freeze the backbone parameters when training starts and unfreeze the end backbone when reaching first milestones and begining when reaching second one.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:42]: can you rephrase? Also, do we have any reference to this technique?

File: docs/source/general/finetuning.rst [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+Use the builts-in Finetuning Strategy Callback.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:44]:
```suggestion
Use the built-in Finetuning Strategy Callback.
```

File: docs/source/general/finetuning.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+Create a custom Finetuning Strategy Callback.
+
+.. code-block:: python
+
+    from flash.core.finetuning import FlashBaseFinetuning
+
+    class FeatureExtractorFreezeUnfreeze(FlashBaseFinetuning):
+
+        def __init__(self, unfreeze_at_epoch: int = 5, train_bn: bool = true)
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:67]:
```suggestion
        def __init__(self, unfreeze_at_epoch: int = 5, train_bn: bool = True)
```

File: docs/source/general/finetuning.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            # this will set self.attr_names as ["feature_extractor"]
+            super().__init__("feature_extractor", train_bn)
+            self._unfreeze_at_epoch = unfreeze_at_epoch
+
+        def freeze_before_training(self, pl_module):
+            # freeze any module you want by overriding this function
+
+            # Here, we are freezing ``feature_extractor``
+            self.freeze_using_attr_names(pl_module, self.attr_names, train_bn=self.train_bn)
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:76]: Not sure about including this function here. I don't want users to think they ned to override it

File: docs/source/general/finetuning.rst (hunk 5 above)
Comment (tchaton) [reply docs/source/general/finetuning.rst:42]: Done ! Not really, but I have seen it several time.

**Review by edenlightning — COMMENTED** [review]
I still think we can be a lot more clear on what freeze and unfreeze means, and when would you like to use which strategy.

File: docs/source/general/finetuning.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
+**********
+Finetuning
+**********
+
+Finetuning (or transfer-learning) is the process of tweaking a model trained on a large dataset, to your particular (likely much smaller) dataset. All Flash tasks have a pre-trained backbone that was trained on large datasets such as ImageNet, and that allows to decrease training time significantly. 
+
+.. note:: If you have a huge dataset and prefer to train from scratch, see the training guide.
```
Comment (Borda) [inline docs/source/general/finetuning.rst:7]:
```suggestion
.. note::

    If you have a huge dataset and prefer to train from scratch, see the training guide.
```

Comment (Borda) [reply docs/source/general/finetuning.rst:7]: btw, you can also use `.. tip::`

File: docs/source/general/finetuning.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+
+Finetune options
+================
+
+Flash supports 4 builts-in Finetuning options:
+
+* `no_freeze`: Don't freeze anything.
+* `freeze`: Freeze the backbone parameters when training starts.
+* `freeze_unfreeze`: Freeze the backbone parameters when training starts and unfreeze the backbone when reaching `unfreeze_epoch`.
```
Comment (Borda) [inline docs/source/general/finetuning.rst:25]: would it work also for spec steps?

File: docs/source/general/finetuning.rst [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+* `unfreeze_milestones`: Freeze the backbone parameters when training starts and unfreeze the end backbone when reaching first milestones and begining when reaching second one.
```
Comment (Borda) [inline docs/source/general/finetuning.rst:26]: can you specify or link what milestone means?

Comment (carmocca) [reply docs/source/general/finetuning.rst:26]: Elaborate on the difference between the "backbone" and the "end backbone".

Also elaborate on what is the beggining

Also "when reaching first milestiones" -> "when the first milestone is reached"

File: docs/source/general/finetuning.rst (hunk 10 above)
Comment (tchaton) [reply docs/source/general/finetuning.rst:36]: Not sure to understand what you mean.

File: docs/source/general/finetuning.rst [hunk 12]
```diff
@@ -0,0 +1,138 @@
+**********
+Finetuning
+**********
+
+Finetuning (or transfer-learning) is the process of tweaking a model trained on a large dataset, to your particular (likely much smaller) dataset. All Flash tasks have a pre-trained backbone that was trained on large datasets such as ImageNet, and that allows to decrease training time significantly.
+
+Finetuning process can be splitted into 4 steps:
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:7]:
```suggestion
The finetuning process can be split into 4 steps:
```

File: docs/source/general/finetuning.rst [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+- 1. Train a source neural network model on a source dataset. For computer vision, it is traditionally  the [ImageNet dataset](http://www.image-net.org/search?q=cat). As training is costly, library such as [Torchvion](https://pytorch.org/docs/stable/torchvision/index.html) library supports popular pre-trainer model architectures.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:9]:
```suggestion
- 1. Train a particular neural network model on a particular dataset. For computer vision, the [ImageNet dataset](http://www.image-net.org/search?q=cat) is widely used. As training is costly, libraries such as [torchvision](https://pytorch.org/docs/stable/torchvision/index.html) provide popular pre-trained model architectures. These are called backbones.
```

File: docs/source/general/finetuning.rst (hunk 10 above)
Comment (Borda) [reply docs/source/general/finetuning.rst:36]: never mind, not important atm :]

File: docs/source/general/finetuning.rst [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+- 2. Create a new neural network called the target model. Its architecture replicates the source model and parameters, expect the latest layer which is removed. This model without its latest layer is traditionally called a backbone
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:11]:
```suggestion
- 2. Create a new neural network called the target model. Its architecture replicates the backbone (model from previous step) and parameters, except the latest layer which is usually replaced to fit the necessities of your data.
```

File: docs/source/general/finetuning.rst [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+- 3. Add new layers after the backbone where the latest output size is the number of target dataset categories. Those new layers, traditionally called head will be randomly initialized while backbone will conserve its pre-trained weights from ImageNet.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:13]:
```suggestion
- 3. This new layer (or layers) at the end of the backbone are used to match the backbone output to the number of target categories in your data. They are commonly referred to as the head'. The head is randomly initialized whereas the backbone conserves its pre-trained weights (for example the weights from ImageNet).
```

File: docs/source/general/finetuning.rst [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+- 4. Train the target model on a smaller target dataset. However, as new layers are randomly initialized, the first gradients will be random when training starts and will destabilize the backbone pre-trained parameters. Therefore, it is good pratice to freeze the backbone, which means the parameters of the backbone won't be trainable for some epochs. After some epochs, the backbone are being unfreezed, meaning the weights will be trainable.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:15]:
```suggestion
- 4. Train the target model on a smaller target dataset. However, as the head (new layers) is untrained, the first results (gradients) will be random when training starts and could decrease the backbone performance (by changing its pre-trained parameters). Therefore, it is a good practice to "freeze" the backbone. This means the parameters of the backbone won't be updated until they are "unfrozen" a few epochs later.
```

File: docs/source/general/finetuning.rst [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+
+.. tip:: If you have a huge dataset and prefer to train from scratch, see the training guide.
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:18]:
```suggestion
.. tip:: If you have a large dataset and prefer to train from scratch, see the training guide.
```

File: docs/source/general/finetuning.rst [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+
+2. Pick a model to run from a variety of Flash tasks: :class:`~flash.vision.ImageClassification`, :class:`~flash.text.TextClassifier`, :class:`~flash.tabular.TabularClassifier`, all optimized with the latest best practices.
+
+3. Finetune your model using  :func:`~flash.Trainer.finetune` method. You will need to choose a finetune strategy.
+
+
+Finetune options
+================
+
+Flash supports 4 builts-in Finetuning options:
```
Comment (carmocca) [inline docs/source/general/finetuning.rst:32]: This is duplicated below
